### PR TITLE
Nix package

### PR DESCRIPTION
**Pull request by SeineEloquenz**
This PR adds a flake with a Nix packaged version of `opp_env`.

To do so, I had to patch `opp_env` in some ways to align the project with the Nix workflow.
The patches are in the `nix` directory in the source tree.

Changes from the included patches:
1. We need to reset the permissions of the `templates` when running `opp_env init`, because when installed via nix, they are copied from the Nix store and therefore are read-only.
2. I removed the pip dependency as well as the upgrade subcommand, as they are not compatible with an installation via Nix.

### CODE REVIEW

**Comment by rhornig** [comment]
Thanks a lot for this. I was also considering publishing `opp_env` automatically to pypi and also creating a proper nix package. Instead of patching the project I would gladly upstream these changes so the opp_env command could behave differently depending on where it was installed from. Currently I see three installation sources:
- using pip from github (this is the only one currently implemented)
- using pip from pypi.org
- using nix

Instead of removing the upgrade subcommand I would prefer to have the same interface in all versions, just giving an error message like `This version of opp_env was installed as a NIX package. Please use "command to upgrade the nix package, etc." instead`

**Comment by SeineEloquenz** [comment]
Thanks for the reply!
I'm not that familiar with pyproject packaging. Can we conditionally specify the pip dependency depending on the build system?
Because if I see it correctly, the dependency is only needed for upgrading and nothing else, so in a nix package we should not have the dependency in the closure at all.

Regarding the `upgrade` subcommand itself I'm gonna look into removing the patches and changing the subcommand accordingly to instead print a message.

**Comment by SeineEloquenz** [comment]
Todos:
[ ]: How to handle the pip dependency
[ ]: How to handle setuptools-scm versioning

I don't know of a good way to handle setuptools-scm. It does not work when building a nix package, because the git metadata is not accessible during build time. One can use `SETUPTOOLS_SCM_PRETEND_VERSION` to still get a built package, but this needs hardcoding the version of the pacakge in the nix file, which currently is a placeholder value to get it to build, which is not exactly a good solution...

**Comment by rhornig** [comment]
How about something like not directly building a nix package, but rather creating a proper pypi package, and releasing it properly. After that as a next step, we could create a nix package that just simply install the opp_env package with pip? I guess a lot of python based tools are packaged like this.

**Comment by SeineEloquenz** [comment]
Sounds good to me. I can work on the Nix package once a pypi release has been done, since I'm not familiar with python packaging.

**Comment by rhornig** [comment]
`opp-env` now automatically builds and deploys to pypi.org. so

    pip install opp-env 

should just work fine.

As a next step, I will review what dependencies we still need. Probably pip and setup tools is no longer needed.

**Comment by rhornig** [comment]
Thanks merged. On longer term, we can enable the package for linux/aarch64 and macos/ (x86/aarch64). I don't have the capabilities to test it right now.

The actual varsion number is baked in the default.nix file which seems a bit error prone. The version numbers are automatically picked by the python build from the git branch tag. I wonder whether we could automatically sync the nix and pypi package versions. I guess this is not so simple as the nix package should not depend on external state (i.e. tags in the git repo). I have to think about this.

Basically we have to ensure that the git tag, the pypi version and the nix version always match (without much manual work).

**Comment by rhornig** [comment]
As long as you have nix, just type: `nix shell github:omnetpp/opp_env` and you get the latest opp_env version in a shell.
